yew/src/virtual_dom/vlist.rs

```diff
@@ -78,100 +85,148 @@ impl VDiff for VList {
         }
 
-        // Process children
-        let lefts = self.children.iter_mut();
-        let key_count = rights.iter().filter(|r| r.key().is_some()).count();
-        let mut rights_nokeys = Vec::with_capacity(rights.len() - key_count);
-        let mut rights_lookup = HashMap::with_capacity(key_count);
-        for mut r in rights.into_iter() {
-            if let Some(key) = r.key() {
-                if rights_lookup.contains_key(key) {
-                    log::error!("Duplicate key of {}", &key);
-                    r.detach(parent);
+        // Collect the keyed elements in a hashmap.
+        let rights_lookup = rights
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, vnode)| vnode.key().as_ref().map(|key| (key.clone(), idx)))
+            .collect::<HashMap<_, _>>();
+
+        // Actual diffing loop.
+        let mut current_right_idx: usize = 0;
+        let mut left_surplus = self.children.len().saturating_sub(rights.len());
+        for left in self.children.iter_mut() {
+            // Ignore the dummy nodes we may have inserted. We can recognize
+            // them because the are VText nodes with empty text and no
+            // reference.
+            while let Some(VNode::VText(vtext)) = rights.get(current_right_idx) {
+                if vtext.reference.is_none() && vtext.text.is_empty() {
+                    current_right_idx += 1;
                 } else {
-                    rights_lookup.insert(key.clone(), r);
+                    break;
                 }
-            } else {
-                rights_nokeys.push(r);
             }
-        }
-        let mut rights = rights_nokeys.into_iter();
-        for left in lefts {
-            if let Some(key) = &left.key() {
-                match rights_lookup.remove(key) {
-                    Some(right) => {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), Some(right));
-                    }
-                    None => {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), None);
-                    }
+
+            let right_key = rights
+                .get(current_right_idx)
+                .and_then(|vnode| vnode.key().as_ref());
+            let (right_idx, reused) = match (left.key(), right_key) {
+                (Some(left_key), Some(right_key)) if left_key == right_key => {
+                    current_right_idx += 1;
+                    (Some(current_right_idx - 1), false)
                 }
-            } else {
-                match rights.next() {
-                    Some(right) => {
-                        previous_sibling =
-                            left.apply(scope, parent, previous_sibling.as_ref(), Some(right));
+                (Some(left_key), _) => match rights_lookup.get(left_key) {
+                    Some(id) if *id > current_right_idx => {
```

Comment:
I think we should match every keyed node in `right` that has a corresponding keyed node in `lefts`. Looks like the current implementation won't match a keyed `right` that is early in the list with a keyed `left` that is late in the list.

Reply:
I wrote a new algorithm.